### PR TITLE
Add new streaming interfaces for the no-wire representation implementation of ThriftRW

### PR DESCRIPTION
**Pull request by dianadkle**
Adds the definitions for the StreamProtocol, Reader, and Writer interfaces, which will be implemented in later PRs.

### CODE REVIEW

File: protocol/protocol.go [hunk 1]
```diff
@@ -71,3 +71,86 @@ type Responder interface {
 	// The EnvelopeType should be either wire.Reply or wire.Exception.
 	EncodeResponse(v wire.Value, t wire.EnvelopeType, w io.Writer) error
 }
+
+// StreamProtocol defines a specific way for a Thrift value to be
+// encoded or decoded, implemented in a streaming fashion.
+type StreamProtocol interface {
+	StreamWriter(io.Writer) Writer
+	StreamReader(io.Reader) Reader
```
Comment (witriew) [inline protocol/protocol.go:79]: Add comments for exported methods.

File: protocol/protocol.go [hunk 2]
```diff
@@ -71,3 +71,86 @@ type Responder interface {
...
+	ReadStructEnd() error
+	ReadFieldBegin() (FieldHeader, bool, error)
+	ReadFieldEnd() error
+	ReadListBegin() (ListHeader, error)
+	ReadListEnd() error
+	ReadSetBegin() (SetHeader, error)
+	ReadSetEnd() error
+	ReadMapBegin() (MapHeader, error)
+	ReadMapEnd() error
+	Skip(wire.Type) error
```
Comment (witriew) [inline protocol/protocol.go:155]: I'm not sure about commenting styles for repeated, patterned methods like above, but `Skip` probably warrants its own comment.

Comment (dianadkle) [reply protocol/protocol.go]: The repeated methods also seemed somewhat self-explanatory, and it felt like any comments that delved into implementation details felt too much into the weeds, though I'm happy to put in comments if that is the convention.

I did add a small comment for skip!

**Review by abhinav — APPROVED** [review]

**Comment by abhinav** [comment]
protocol/protocol.go is failing `gofmt`. Please run gofmt on it.

**Review by witriew — APPROVED** [review]

**Pull request merged**